src/Nix/NixString.hs

```diff
@@ -0,0 +1,68 @@
+{-# LANGUAGE DeriveGeneric #-}
+module Nix.NixString (
+    NixString
+  , hackyStringHasContext
+  , hackyStringIgnoreContextMaybe
+  , hackyStringIgnoreContext
+  , makeNixStringWithoutContext
+  , modifyNixContents
+) where
+
+import qualified Data.HashSet as S
+import           Data.Hashable
+import           Data.Text (Text)
+import           GHC.Generics
+import           Data.Semigroup
+
+{-# WARNING hackyStringHasContext, hackyStringIgnoreContextMaybe, hackyStringIgnoreContext "This NixString function needs to be replaced" #-}
+
+-- | A 'ContextFlavor' describes the sum of possible derivations for string contexts
+data ContextFlavor = 
+    DirectPath
+  | DerivationOutput !Text
+  deriving (Show, Eq, Ord, Generic)
+
+instance Hashable ContextFlavor
+
+-- | A 'StringContext' ...
+data StringContext = 
+  StringContext { scPath :: !Text
+                , scFlavor :: !ContextFlavor
+                } deriving (Eq, Ord, Show, Generic) 
+
+instance Hashable StringContext
+
+data NixString = NixString 
+  { nsContents :: !Text
+  , nsContext :: !(S.HashSet StringContext)
+  } deriving (Eq, Ord, Show, Generic) 
+
+instance Hashable NixString
+
+instance Semigroup NixString where
+  NixString s1 t1 <> NixString s2 t2 = NixString (s1 <> s2) (t1 <> t2)
```

Comment:
Let's leave this for now, but I am a tiny bit concerned that it might let context mishandling sneak in somewhere. Just to flag it in case we're seeing weird bugs in the future :joy: